cinn/optim/vectorize_loops.cc

```diff
@@ -50,6 +55,177 @@ Expr Widen(Expr e, int lanes) {
   return ir::Broadcast::Make(e, lanes);
 }
 
+// tell whether a tensor can be vectorized or not on CUDA by collecting names
+// of tensors which doesn't meet all predicates of vectoring
+class TensorVectorizeExcluder : public ir::IRMutator<const Expr *> {
+ public:
+  TensorVectorizeExcluder(const std::string &iter_name) : iter_name_(iter_name) {}
+
+  void Collect(const Expr *op) { IRMutator::Visit(op, op); }
+
+  bool Contain(const std::string &tensor_name) const { return excluded_tensors_.count(tensor_name); }
+
+ private:
+  const std::string iter_name_;
+  std::set<std::string> excluded_tensors_;
+
+  void Visit(const ir::Store *expr, const Expr *op) override {
+    auto *node = op->As<ir::Store>();
+    CHECK(node);
+    IRMutator::Visit(&node->value, &node->value);
+    auto *tensor = node->tensor.As<ir::_Tensor_>();
+    CHECK(tensor);
+    if (HitBan(node->tensor, node->indices)) excluded_tensors_.insert(tensor->name);
+  }
+
+  void Visit(const ir::Load *expr, const Expr *op) override {
+    auto *node = op->As<ir::Load>();
+    CHECK(node);
+    auto *tensor = node->tensor.As<ir::_Tensor_>();
+    CHECK(tensor);
+    if (HitBan(node->tensor, node->indices)) excluded_tensors_.insert(tensor->name);
+  }
+
+  // return true if the tensor hit some prohibitions of vectorizing
+  bool HitBan(Expr tensor, const std::vector<Expr> &indices) {
+    auto find_matched_var_fn = [&](const Expr *x) { return x->As<_Var_>() && x->As<_Var_>()->name == iter_name_; };
+    auto in_last_index       = ir::CollectIRNodes(indices.back(), find_matched_var_fn);
+
+    // the iter val must appear in the last index
+    if (in_last_index.empty()) {
+      VLOG(5) << "Loop var:" << iter_name_
+              << " is not used in the last index, so the memory access will be not sequential";
+      return true;
+    }
+
+    // the iter val can't appear in mulitple indices
+    for (int i = 0; i < indices.size() - 1; ++i) {
```

Comment:
Is it possible that `indices.size()` is 0? If yes, this code can be dangerous, since `indices.size()` is `size_t` data type, not `int`, `indices.size() - 1` would be very big number and cause bug.

Can we check the pre-condition of indices.size()?

Reply:
add an extra check whether the indices is empty and will quit immediately if it is